File: app/src/main/java/org/neidhardt/dynamicsoundboard/navigationdrawer/playlist/Playlist.java
package org.neidhardt.dynamicsoundboard.navigationdrawer.playlist;

import android.content.Context;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import de.greenrobot.event.EventBus;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.navigationdrawer.NavigationDrawerList;
import org.neidhardt.dynamicsoundboard.navigationdrawer.events.PlaylistSoundsRemovedEvent;
import org.neidhardt.dynamicsoundboard.views.DividerItemDecoration;

import java.util.ArrayList;
import java.util.List;

/**
 * Project created by eric.neidhardt on 27.08.2014.
 */
public class Playlist extends NavigationDrawerList implements PlaylistAdapter.OnItemClickListener
{
	public static final String TAG = Playlist.class.getName();

	private RecyclerView playlist;
	private PlaylistAdapter adapter;

	@SuppressWarnings("unused")
	public Playlist(Context context)
	{
		super(context);
		this.inflateLayout(context);
		this.initRecycleView(context);
	}

	@SuppressWarnings("unused")
	public Playlist(Context context, AttributeSet attrs)
	{
		super(context, attrs);
		this.inflateLayout(context);
		this.initRecycleView(context);
	}

	@SuppressWarnings("unused")
	public Playlist(Context context, AttributeSet attrs, int defStyle)
	{
		super(context, attrs, defStyle);
		this.inflateLayout(context);
		this.initRecycleView(context);
	}

	private void inflateLayout(Context context)
	{
		LayoutInflater.from(context).inflate(R.layout.view_playlist, this, true);
	}

	private void initRecycleView(Context context)
	{
		this.playlist = (RecyclerView)this.findViewById(R.id.rv_playlist);
		if (!this.isInEditMode())
		{
			this.playlist.addItemDecoration(new DividerItemDecoration());
			this.playlist.setLayoutManager(new LinearLayoutManager(context));
			this.playlist.setItemAnimator(new DefaultItemAnimator());
		}
	}

	@Override
	protected void onAttachedToWindow()
	{
		super.onAttachedToWindow();
		EventBus bus = EventBus.getDefault();
		if (!bus.isRegistered(this.adapter))
			bus.register(this.adapter);

		this.adapter.notifyDataSetChanged();
	}

	@Override
	protected void onDetachedFromWindow()
	{
		EventBus.getDefault().unregister(this.adapter);
		super.onDetachedFromWindow();
	}

	public void setAdapter(PlaylistAdapter adapter)
	{
		this.adapter = adapter;
		this.adapter.setOnItemClickListener(this);
		this.playlist.setAdapter(adapter);
		this.adapter.setRecyclerView(this.playlist);
	}

	@Override
	public void onItemClick(View view, EnhancedMediaPlayer player, int position)
	{
		if (super.isInSelectionMode)
			super.onItemSelected(view, position);
		else
			this.adapter.startOrStopPlayList(player);
	}

	@Override
	protected int getActionModeTitle()
	{
		return R.string.cab_title_delete_play_list_sounds;
	}

	@Override
	protected void onDeleteSelected(SparseArray<View> selectedItems)
	{
		List<EnhancedMediaPlayer> playersToRemove = new ArrayList<>(selectedItems.size());
		for(int i = 0; i < selectedItems.size(); i++)
		{
			int index = selectedItems.keyAt(i);
			playersToRemove.add(this.adapter.getValues().get(index));
		}

		EventBus.getDefault().post(new PlaylistSoundsRemovedEvent(playersToRemove));
		this.adapter.notifyDataSetChanged();
	}

	@Override
	protected int getItemCount() {
		return this.adapter.getItemCount();
	}

	public void notifyDataSetChanged()
	{
		this.adapter.notifyDataSetChanged();
	}

}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/navigationdrawer/playlist/Playlist.kt
package org.neidhardt.dynamicsoundboard.navigationdrawer.playlist

import android.content.Context
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.LayoutInflater
import org.greenrobot.eventbus.EventBus
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.navigationdrawer.views.NavigationDrawerList
import org.neidhardt.dynamicsoundboard.views.recyclerviewhelpers.DividerItemDecoration

/**
 * Project created by eric.neidhardt on 27.08.2014.
 */
class Playlist : NavigationDrawerList
{
	companion object
	{
		val TAG: String = Playlist::class.java.name
	}

	private val soundsDataStorage = SoundboardApplication.getSoundsDataStorage()
	private val soundsDataAccess = SoundboardApplication.getSoundsDataAccess()

	val presenter: PlaylistPresenter = PlaylistPresenter(EventBus.getDefault(), this.soundsDataStorage, this.soundsDataAccess)
	val adapter:PlaylistAdapter = PlaylistAdapter(this.presenter)

	@SuppressWarnings("unused") constructor(context: Context) : super(context)
	{
		this.init(context)
	}

	@SuppressWarnings("unused") constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
	{
		this.init(context)
	}

	@SuppressWarnings("unused") constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)
	{
		this.init(context)
	}

	private fun init(context: Context)
	{
		this.presenter.adapter = this.adapter

		LayoutInflater.from(context).inflate(R.layout.view_playlist, this, true)

		val playlist = this.findViewById(R.id.rv_playlist) as RecyclerView
		if (!this.isInEditMode) {
			playlist.addItemDecoration(DividerItemDecoration(this.context))
			playlist.layoutManager = LinearLayoutManager(this.context)
			playlist.itemAnimator = DefaultItemAnimator()
		}
		playlist.adapter = this.adapter

		this.adapter.recyclerView = playlist
	}

	override fun onAttachedToWindow()
	{
		super.onAttachedToWindow()
		this.presenter.onAttachedToWindow()
	}

	override fun onDetachedFromWindow()
	{
		this.presenter.onDetachedFromWindow()
		super.onDetachedFromWindow()
	}

	override fun onFinishInflate()
	{
		super.onFinishInflate()
		this.presenter.view = this
	}

	override val itemCount: Int
		get() = this.adapter.itemCount

	override val actionModeTitle: Int
		get() = R.string.cab_title_delete_play_list_sounds

}
